core/src/main/java/org/elasticsearch/index/mapper/MapperService.java

```diff
@@ -301,9 +337,37 @@ private boolean assertSerialization(DocumentMapper mapper) {
         return true;
     }
 
-    protected void checkNewMappersCompatibility(Collection<ObjectMapper> newObjectMappers, Collection<FieldMapper> newFieldMappers, boolean updateAllTypes) {
+    /**
+     * Check that fields are defined only once. It is possible to define fields
+     * several times eg. if a field is defined once via 'fields' in the mapping
+     * and once from a field mapper.
+     */
+    private void checkUniqueness(Collection<ObjectMapper> objectMappers, Collection<FieldMapper> fieldMappers) {
+        Map<String, Mapper> alreadySeen = new HashMap<>();
+        for (ObjectMapper mapper : objectMappers) {
+            final Mapper removed = alreadySeen.put(mapper.fullPath(), mapper);
+            if (removed != null) {
+                throw new IllegalArgumentException("Two fields are defined for path [" + mapper.fullPath() + "]: " + Arrays.asList(mapper, removed));
+            }
+        }
+        for (FieldMapper mapper : fieldMappers) {
+            final Mapper removed = alreadySeen.put(mapper.name(), mapper);
+            if (removed != null && (indexSettings.getIndexVersionCreated().onOrAfter(Version.V_2_0_0) || mapper != removed)) {
```

Comment:
Can we fix this when upgrading the mappings, so we don't have to keep this leniency in the check?